File: frontend/src/types/invoice.ts
export interface Invoice {
  id: string;
  invoice_number: string;
  contract_id: string;
  unit_id: string;
  unit_number: string;
  renter_name: string;
  billing_period_start: string;
  billing_period_end: string;
  due_date: string;
  subtotal: number;
  total: number;
  status: 'draft' | 'sent' | 'paid' | 'overdue' | 'cancelled';
  paid_amount: number;
  paid_at?: string;
  payment_method?: string;
  vietqr_ref_code?: string;
  notes?: string;
  created_at: string;
}

export interface InvoiceDetail extends Invoice {
  items: InvoiceItem[];
  qr_image_base64?: string;
}

export interface InvoiceItem {
  id: string;
  item_type: string;
  description: string;
  quantity: number;
  unit_price: number;
  amount: number;
}

export interface InvoiceFilter {
  status?: string;
  unit_id?: string;
  renter_id?: string;
  billing_month?: string;
}

export interface InvoiceItemCreate {
  item_type: 'rent' | 'electricity' | 'water' | 'service' | 'internet' | 'parking' | 'other';
  description: string;
  quantity?: number;
  unit_price?: number;
  amount: number;
}

export interface InvoiceCreate {
  contract_id: string;
  billing_period_start: string;
  billing_period_end: string;
  due_date: string;
  items: InvoiceItemCreate[];
  notes?: string;
}

export interface MarkPaidRequest {
  payment_method: string;
  paid_amount: number;
}

File: backend/app/models/invoice.py
from pydantic import BaseModel
from typing import Optional, Literal


class InvoiceItemCreate(BaseModel):
    item_type: Literal['rent', 'electricity', 'water', 'service', 'internet', 'parking', 'other']
    description: str
    quantity: Optional[float] = 1
    unit_price: Optional[float] = 0
    amount: float


class InvoiceCreate(BaseModel):
    contract_id: str
    billing_period_start: str
    billing_period_end: str
    due_date: str
    items: list[InvoiceItemCreate]
    notes: Optional[str] = None


class MarkPaidRequest(BaseModel):
    payment_method: str
    paid_amount: float


class AutoGenerateRequest(BaseModel):
    billing_month: Optional[str] = None


class InvoiceResponse(BaseModel):
    id: str
    invoice_number: str
    contract_id: str
    unit_id: Optional[str] = None
    unit_number: Optional[str] = None
    renter_name: Optional[str] = None
    billing_period_start: str
    billing_period_end: str
    due_date: str
    subtotal: float
    total: float
    status: str
    paid_amount: float
    paid_at: Optional[str] = None
    payment_method: Optional[str] = None
    vietqr_ref_code: Optional[str] = None
    notes: Optional[str] = None
    created_at: Optional[str] = None


class InvoiceItemResponse(BaseModel):
    id: str
    item_type: str
    description: str
    quantity: float
    unit_price: float
    amount: float


class InvoiceDetailResponse(InvoiceResponse):
    items: list[InvoiceItemResponse] = []
    qr_image_base64: Optional[str] = None


class PaymentWebhookPayload(BaseModel):
    id: Optional[int] = None
    tid: Optional[str] = None
    description: str
    amount: float
    cusum_balance: Optional[float] = None
    when: Optional[str] = None
    bank_sub_acc_id: Optional[str] = None
